consensus/src/ledger/indexed_merkle_tree.rs

```diff
@@ -0,0 +1,129 @@
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{marker::PhantomData, sync::Arc};
+
+use anyhow::*;
+use indexmap::IndexSet;
+use snarkos_storage::Digest;
+use snarkvm_algorithms::{merkle_tree::MerkleTree, MerkleParameters};
+use snarkvm_utilities::ToBytes;
+
+pub struct IndexedMerkleTree<P: MerkleParameters> {
+    tree: MerkleTree<P>,
+    index: IndexSet<Digest>,
+}
+
+impl<P: MerkleParameters> Clone for IndexedMerkleTree<P> {
+    fn clone(&self) -> Self {
+        let tree = self
+            .tree
+            .rebuild::<[u8; 32], _>(DummyLength(self.index.len(), PhantomData), &[])
+            .expect("failed to clone merkle tree");
+        Self {
+            tree,
+            index: self.index.clone(),
+        }
+    }
+}
+
+struct DummyLength<T>(usize, PhantomData<T>);
+
+impl<T> Iterator for DummyLength<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        panic!("attempted to iterate on dummy length iterator");
+    }
+}
+
+impl<T> ExactSizeIterator for DummyLength<T> {
+    fn len(&self) -> usize {
+        self.0
+    }
+}
+
+fn to_digest<B: ToBytes>(input: &B) -> Result<Digest> {
+    let mut data = vec![];
+    input.write(&mut data)?;
+    Ok((&data[..]).into())
+}
+
+impl<P: MerkleParameters> IndexedMerkleTree<P> {
+    pub fn new(parameters: Arc<P>, leaves: &[Digest]) -> Result<Self> {
+        Ok(Self {
+            tree: MerkleTree::new(parameters, leaves)?,
+            index: leaves.iter().cloned().collect(),
+        })
+    }
+
+    pub fn extend(&mut self, new_leaves: &[Digest]) -> Result<()> {
+        let tree = self
+            .tree
+            .rebuild(DummyLength(self.index.len(), PhantomData), new_leaves)?;
+        self.tree = tree;
+        self.index.extend(new_leaves.iter().cloned());
+        Ok(())
+    }
+
+    /// pop leafs from the interior merkle tree, and assert they are equal to `to_remove`.
+    pub fn pop(&mut self, to_remove: &[Digest]) -> Result<()> {
+        if to_remove.len() > self.index.len() {
+            return Err(anyhow!(
+                "attempted to remove more items from indexed merkle tree than present"
+            ));
+        }
+        let old_length = self.index.len() - to_remove.len();
+        for i in old_length..self.index.len() {
+            if self.index[i] != to_remove[i] {
```

Comment:
shouldn't `to_remove` be indexed starting from `0`?

Reply:
yep.